nixos/modules/security/systemd-confinement.nix

```diff
@@ -0,0 +1,192 @@
+            <manvolnum>2</manvolnum>
+          </citerefentry>.
+        '';
+      };
+
+      options.confinement.fullUnit = lib.mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to include the full closure of the systemd unit file into the
+          chroot, instead of just the dependencies for the executables.
+
+          <warning><para>While it may be tempting to just enable this option to
+          make things work quickly, please be aware that this might add paths
+          to the closure of the chroot that you didn't anticipate. It's better
+          to use <option>confinement.packages</option> to <emphasis
+          role="strong">explicitly</emphasis> add additional store paths to the
+          chroot.</para></warning>
+        '';
+      };
+
+      options.confinement.packages = lib.mkOption {
+        type = types.listOf (types.either types.str types.package);
+        default = [];
+        description = let
+          mkScOption = optName: "<option>serviceConfig.${optName}</option>";
+        in ''
+          Additional packages or strings with context to add to the closure of
+          the chroot. By default, this includes all the packages from the
+          ${lib.concatMapStringsSep ", " mkScOption [
+            "ExecReload" "ExecStartPost" "ExecStartPre" "ExecStop"
+            "ExecStopPost"
+          ]} and ${mkScOption "ExecStart"} options. If you want to have all the
+          dependencies of this systemd unit, you can use
+          <option>confinement.fullUnit</option>.
+
+          <note><para><emphasis role="strong">Only</emphasis> the latter
+          (${mkScOption "ExecStart"}) will be used if
+          ${mkScOption "RootDirectoryStartOnly"} is enabled.</para></note>
+
+          <note><para>Also, the store paths listed in <option>path</option> are
+          <emphasis role="strong">not</emphasis> included in the closure as
+          well as paths from other options except those listed
+          above.</para></note>
+        '';
+      };
+
+      options.confinement.binSh = lib.mkOption {
+        type = types.nullOr types.path;
+        default = toplevelConfig.environment.binsh;
+        defaultText = "config.environment.binsh";
+        example = lib.literalExample "\${pkgs.dash}/bin/dash";
+        description = ''
+          The program to make available as <filename>/bin/sh</filename> inside
+          the chroot. If this is set to <literal>null</literal>, no
+          <filename>/bin/sh</filename> is provided at all.
+
+          This is useful for some applications, which for example use the
+          <citerefentry>
+            <refentrytitle>system</refentrytitle>
+            <manvolnum>3</manvolnum>
+          </citerefentry> library function to execute commands.
+        '';
+      };
+
+      options.confinement.mode = lib.mkOption {
+        type = types.enum [ "full-apivfs" "chroot-only" ];
```

Comment:
What's this option useful for? If people want to turn off certain isolation options, they can set `PrivateDevices = false`, etc. themselves. This option would only introduce one additional layer of abstraction, of which NixOS has enough already imo

Reply:
This option is actually very useful, particularly the `chroot-only` mode, which allows the application to take care of further restricting itself on its own. Not needing to have `/dev`, `/proc` and `/sys` lowers the attack surface by a huge factor, but of course the application needs to have sandboxing built-in (LibreOffice Online does that for example).

Reply:
Yeah, but this all is still possible if we remove this option and have a `full-apivfs` setup by default (setting all the `serviceConfig` sandboxy values to `true` with a `mkDefault`), and then let the service writers turn off every protection they don't want manually.

Reply:
I beg to differ here, because the unit options picked by `full-apivfs` unfortunately quite version-specific to systemd. For example I would rather _avoid_ the need to use the `Protect*` options, because they actually increase the attack surface (because they use a blacklist-approach) but right now there is no better option as you just can enable APIVFS or disable it, rather than specifying which pseudo-fs to mount. Additionally, I also want to get in PID namespacing as soon as systemd provides an option.

That's why the `chroot-only` option gets you to the point where you don't need to worry about future systemd updates. Let's say you only want a chroot without anything else and PID namespacing gets enabled, you need to keep track of that and disable it accordingly (after you're done debugging why your service suddenly fails).

Reply:
Alright this sounds more reasonable then.